pkg/service/controller.go

```diff
@@ -850,3 +891,55 @@ func (c *ControllerService) ControllerGetCapabilities(context.Context, *csi.Cont
 func (c *ControllerService) ControllerGetVolume(_ context.Context, _ *csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
+// IsVolumeAttachedToOtherVMI checks if a PVC is actively
+// used by any VirtualMachineInstance other than the current one.
+//
+// NOTE: This function uses vmi.Status.VolumeStatus as the source of truth for
+// what is currently attached. It directly compares the volume name in the status
+// with the target PVC name.
+//
+// Parameters:
+//   - ctx: The context for cancellation.
+//   - dvName: The name of the PersistentVolumeClaim to check for.
+//   - infraNamespace: The namespace of the PersistentVolumeClaim.
+//   - currentVMIName: The name of the VMI for the current ControllerPublishVolume
+//     request. We want to ignore this VMI in our check.
+//
+// Returns:
+//   - bool: True if the volume is attached to another VMI.
+//   - error: An error if listing VMIs fails.
+func (c *ControllerService) IsVolumeAttachedToOtherVMI(
+	ctx context.Context,
+	dvName string,
+	infraNamespace string,
+	currentVMIName string,
+) (bool, error) {
+	vmis, err := c.virtClient.ListVirtualMachines(ctx, infraNamespace)
+	if err != nil {
+		return false, fmt.Errorf("failed to list Virtual Machine Instances in namespace %s: %w", infraNamespace, err)
+	}
+
+	for _, vmi := range vmis {
+		// Skip the VMI that the volume is intended for.
+		if vmi.Name == currentVMIName {
+			continue
+		}
+
+		// The source of truth is the VMI's status. We iterate through the volumes
+		// that are reported as active in the status.
+		for _, volumeStatus := range vmi.Status.VolumeStatus {
+			// If the name in the status matches our PVC name, it means the volume
+			// is actively attached to this other VMI.
+			if volumeStatus.Name == dvName {
```

Comment:
Do we want to be a little more explicit. The volume could be in the process of unplugging, or won't that matter in this case?